### ISSUE TITLE
Titanic: Machine Learning from Disaster --GBDT get the best

### ISSUE THREAD

**Issue by DYnamic0617** [DYnamic0617/kaggle_playground#1]
The sinking of the RMS Titanic is one of the most infamous shipwrecks in history.  On April 15, 1912, during her maiden voyage, the Titanic sank after colliding with an iceberg, killing 1502 out of 2224 passengers and crew. This sensational tragedy shocked the international community and led to better safety regulations for ships.

One of the reasons that the shipwreck led to such loss of life was that there were not enough lifeboats for the passengers and crew. Although there was some element of luck involved in surviving the sinking, some groups of people were more likely to survive than others, such as women, children, and the upper-class.

In this challenge, we ask you to complete the analysis of what sorts of people were likely to survive. In particular, we ask you to apply the tools of machine learning to predict which passengers survived the tragedy.

`#导入pandas方便数据读取和预处理。
import pandas as pd

#分别对训练和测试数据从本地进行读取。
train = pd.read_csv('C:/Program Files (x86)/python/Datasets/titanic/train.csv')
test = pd.read_csv('C:/Program Files (x86)/python/Datasets/titanic/test.csv')

#先分别输出训练与测试数据的基本信息。这是一个好习惯，可以对数据的规模、各个特征的数据类型以及是否有缺失等，有一个总体的了解。
#print(train.info())
#print(test.info())

#按照之前对Titanic事件的经验，人工选取对预测有效的特征。
selected_features = ['Pclass', 'Sex', 'Age', 'Embarked', 'SibSp', 'Parch', 'Fare']
X_train = train[selected_features]
X_test = test[selected_features]
y_train = train['Survived']

#通过之前的总体观察，得知embarked特征存在缺失值，需要补全--特征处理
#print(X_train['Embarked'].value_counts())
#print(X_test['Embarked'].value_counts())

#对于Embarked这种类别类型的特征，使用出现频率最高的特征值来填充，这也是相对减少引入误差的一种填充方法。
X_train['Embarked'].fillna('S', inplace=True)
X_test['Embarked'].fillna('S', inplace=True)

#而对于age这种数值类型的特征，使用平均值或者中位数填充缺失值，可以相对减少引入误差的一种填充方法。
X_train['Age'].fillna(X_train['Age'].mean(), inplace=True)
X_test['Age'].fillna(X_train['Age'].mean(), inplace=True)

X_test['Fare'].fillna(X_test['Fare'].mean(), inplace=True)

#重新对处理后的训练和测试数据进行查检，发现一切就绪。
#print(X_train.info())
#print(X_test.info())

#从sklearn.feature_extraction导入DictVectorizer。
from sklearn.feature_extraction import DictVectorizer
vec = DictVectorizer(sparse=False)

#对原数据进行特征向量化处理。
X_train = vec.fit_transform(X_train.to_dict(orient='record'))
X_test = vec.transform(X_test.to_dict(orient='record'))

#print(vec.feature_names_)

#从sklearn.ensemble中导入RandomForestClassifier。
from sklearn.ensemble import RandomForestClassifier
#使用默认配置初始化RandomForestClassifier。
rfc = RandomForestClassifier()

#从流行工具包xgboost导入XGBClassifier用于处理分类预测问题。
from xgboost import XGBClassifier
#也使用默认配置初始化XGBClassifier。
xgbc = XGBClassifier()

#从sklearn.ensemble中导入GradientBoostingClassifier。
from sklearn.ensemble import GradientBoostingClassifier
#使用默认配置初始化GradientBoostingClassifier。
gbc = GradientBoostingClassifier()

from sklearn.cross_validation import cross_val_score
#使用5折交叉验证的方法在训练集上分别对默认配置的RandomForestClassifier以及XGBClassifier进行性能评估，并获得平均分类准确性的得分。
cross_val_score(rfc, X_train, y_train, cv=5).mean()
cross_val_score(xgbc, X_train, y_train, cv=5).mean()
cross_val_score(gbc, X_train, y_train, cv=5).mean()

#使用默认配置的RandomForestClassifier进行预测操作。
rfc.fit(X_train, y_train)
rfc_y_predict = rfc.predict(X_test)
rfc_submission = pd.DataFrame({'PassengerId': test['PassengerId'], 'Survived': rfc_y_predict})

#将默认配置的RandomForestClassifier对测试数据的预测结果存储在文件rfc_submission.csv中。
rfc_submission.to_csv('C:/Program Files (x86)/python/Datasets/titanic/rfc_submission.csv', index=False)

#使用默认设置的XGBClassifier进行预测操作。
xgbc.fit(X_train, y_train)
xgbc_y_predict = xgbc.predict(X_test)
xgbc_submission = pd.DataFrame({'PassengerId': test['PassengerId'], 'Survived': xgbc_y_predict})
xgbc_submission.to_csv('C:/Program Files (x86)/python/Datasets/titanic/xgbc_submission.csv', index=False)

#使用默认设置的GradientBoostingClassifier进行预测操作。
gbc.fit(X_train, y_train)
gbc_y_predict = gbc.predict(X_test)
gbc_submission = pd.DataFrame({'PassengerId': test['PassengerId'], 'Survived': gbc_y_predict})
gbc_submission.to_csv('C:/Program Files (x86)/python/Datasets/titanic/gbc_submission.csv', index=False)
`
